prow/cmd/deck/main.go

```diff
@@ -1468,12 +1485,203 @@ func canTriggerJob(user string, pj prowapi.ProwJob, cfg *prowapi.RerunAuthConfig
+	}
+
+	prowJobSpec := pjutil.PresubmitSpec(*presubmitJob, *refs)
+	return &prowJobSpec, presubmitJob.Labels, presubmitJob.Annotations, nil
+}
+
+type postsubmitJobHandler struct{}
+
+func (poh *postsubmitJobHandler) getProwJobSpec(cfg config.Getter, pe ProwJobEvent) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
+	// postsubmit jobs require Refs to be set
+	refs := pe.Refs
+	if refs == nil {
+		return nil, nil, nil, errors.New("refs must be supplied")
+	}
+	if len(refs.Org) == 0 {
+		return nil, nil, nil, errors.New("org must be supplied")
+	}
+	if len(refs.Repo) == 0 {
+		return nil, nil, nil, errors.New("repo must be supplied")
+	}
+	if len(refs.BaseSHA) == 0 {
+		return nil, nil, nil, errors.New("baseSHA must be supplied")
+	}
+	if len(refs.BaseRef) == 0 {
+		return nil, nil, nil, errors.New("baseRef must be supplied")
+	}
+
+	var postsubmitJob *config.Postsubmit
+	org, repo, branch := refs.Org, refs.Repo, refs.BaseRef
+	orgRepo := org + "/" + repo
+
+	postsubmits := cfg().GetPostsubmitsStatic(orgRepo)
+
+	for _, job := range postsubmits {
+		job := job
+		if !job.CouldRun(branch) { // filter out jobs that are not branch matching
+			continue
+		}
+		if job.Name == pe.Name {
+			if postsubmitJob != nil {
+				return nil, nil, nil, fmt.Errorf("%s matches multiple prow jobs from orgRepo %q", pe.Name, orgRepo)
+			}
+			postsubmitJob = &job
+		}
+	}
+	if postsubmitJob == nil {
+		return nil, nil, nil, fmt.Errorf("failed to find associated postsubmit job %q from orgRepo %q", pe.Name, orgRepo)
+	}
+
+	prowJobSpec := pjutil.PostsubmitSpec(*postsubmitJob, *refs)
+	return &prowJobSpec, postsubmitJob.Labels, postsubmitJob.Annotations, nil
+}
+
+type periodicJobHandler struct{}
+
+func (peh *periodicJobHandler) getProwJobSpec(cfg config.Getter, pe ProwJobEvent) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
+	var periodicJob *config.Periodic
+	for _, job := range cfg().AllPeriodics() {
+		if job.Name == pe.Name {
+			// Directly followed by break, so this is ok
+			// nolint: exportloopref
+			periodicJob = &job
+			break
+		}
+	}
+	if periodicJob == nil {
+		return nil, nil, nil, fmt.Errorf("failed to find associated periodic job %q", pe.Name)
+	}
+
+	prowJobSpec := pjutil.PeriodicSpec(*periodicJob)
+	return &prowJobSpec, periodicJob.Labels, periodicJob.Annotations, nil
+}
+
+// Valid value for query parameter mode in rerun route
+const (
+	LATEST = "latest"
+)
+
+func newRerunProwJob(mode string, cfg config.Getter, pj *v1.ProwJob) (v1.ProwJob, error) {
+	if mode == LATEST {
+		pe := ProwJobEvent{
+			Name: pj.Spec.Job,
+			Refs: pj.Spec.Refs,
+		}
+
+		var jh jobHandler
+		switch pj.Spec.Type {
+		case prowapi.PresubmitJob:
+			jh = &presubmitJobHandler{}
+		case prowapi.PostsubmitJob:
+			jh = &postsubmitJobHandler{}
+		case prowapi.PeriodicJob:
+			jh = &periodicJobHandler{}
+		default:
+			return v1.ProwJob{}, fmt.Errorf("Invalid prowjob type")
```

Comment:
there is no formatting, should use `errors.New()` instead

Reply:
not used anymore!